_posts/2021-03-17-encoding-effects-with-freer-simple.markdown

```diff
@@ -0,0 +1,474 @@
+toServantError (Forbidden msg)     = throwError err403 { errBody = toSL msg }
+{% endhighlight %}
+
+Then we have our *Program* and our *Interpretation*, therefore it is a matter of composing them:
+
+{% highlight haskell %}
+run :: Conf -> Document -> Handler DocRef
+run conf = runEffects conf . uploadDocument
+{% endhighlight %}
+
+## Test
+
+As we are going to see now, implementing **Mocks** and Testing our *Effectful Programs* in a deterministic manner is straightforward and easy.
+
+{% highlight haskell %}
+
+data Boundaries = Boundaries
+  { _bUserExists      :: Bool
+  , _bFailSetUploaded :: Bool
+  , _bFailUpload      :: Bool
+  , _bDownload        :: DownloadBoundary
+  } deriving (Generic)
+
+runTest :: Eff '[DocStorage, DataAccess, Logging, AppError, Identity] a -> Either Err a
+runTest = runWithTest def
+
+runWithTest :: Boundaries -> Eff '[DocStorage, DataAccess, Logging, AppError, Identity] a -> Either Err a
+runWithTest boundaries = runIdentity . runM . runError . runTestColog . runPostgresqlTest boundaries . runS3StorageTest boundaries
+
+runTestColog :: forall effs a. Eff (Logging ': effs) a 
+             -> Eff effs a
+runTestColog req = do
+    ((x, _), _) <- F.reinterpret2 go req & FS.runState [] & runWriter
+    pure x
+  where
+    go :: Logging v
+       -> Eff (FS.State [Text] ': FW.Writer [Text] ': effs) v
+    go (LogInfo msg)  = tell ["INFO: "<>msg]
+    go (LogError msg) = tell ["ERROR: "<>msg]
+
+runPostgresqlTest :: Members '[ AppError ] effs => Boundaries -> Eff (DataAccess ': effs) a  -> Eff effs a
+runPostgresqlTest Boundaries{..} = 
+    interpret $ 
+        \case
+            UserExists userId -> return _bUserExists
+            SetDocumentUploaded _ _  -> when _bFailSetUploaded $ FE.throwError ERROR_UPLOADING_DOCUMENT
+
+
+runS3StorageTest :: Members '[ AppError ] effs => Boundaries -> Eff (DocStorage ': effs) a -> Eff effs a
+runS3StorageTest Boundaries{..} =
+    interpret $ 
+        \case
+            UploadDocument _ -> if _bFailUpload 
+                                  then FE.throwError ERROR_UPLOADING_DOCUMENT
+                                  else return $ DocRef "test_ref"
+
+{% endhighlight %}
+
+It is just a matter of implementing the right interpreters in the format that we want that allow us to run any test or combination we might need.
+
+## Conclusions
+**freer-simple** library is an easy **Effect System** library that is based on Academic Research Papers and provide a Robust and Extensible way to build *Effectful Programs*.
+On the other hand we have seen the flexibility this tooling has providing us with some built-in *effects* like `Reader`, `Writer`, `State` and so on.
+Finally we have seen how we can quickly define *pure* interpreters or *Handlers* to escape from the *non-deterministic* context and be able to Test our programs. This interpretations are easy to built, does not require multiple `newtype` declarations or **Monad** mock instances and moreover are highly composable.
```

Comment:
"does not require multiple `newtype` declarations or **Monad** mock instances"

I would remove this part of the sentence since it is confusing. 
Maybe you can say that it solve the n^2 problem of mtl instances.
And you don't need to mock monad instances but still you need to add a mock interpreter.